TheJumpCloud/support
# State param implementation

## Issues
* [SA-2373](https://jumpcloud.atlassian.net/browse/SA-2373) - Update New/Set/Get JCUser to include state parameters

## What does this solve?
The `-suspended` boolean parameter is going to be depreciated in favor of the `-state` parameter. Admins will be able to set the user's state with the `-state` parameter in accordance to the workflow mentioned in the [Managing User States](https://support.jumpcloud.com/support/s/article/Managing-User-States) help article

## Is there anything particularly tricky?
When piping in an object, there was an issue using `Set-JcUser` where the function would reference the parameter set from the object rather than the input from the user. A workaround was to set `ValueFromPipelineByPropertyName` to `$false`, which will ignore the value of the suspended param from the object and instead reference the user's input. 

We still want to maintain functionality of the `-suspended` parameter for the time being to minimize customer impact

## How should this be tested?
_NOTE: The tests below have been added to the CI pipeline for automated testing_
1. Pull the branch to local machine
2. Import the module into a new session 
`Import-Module “/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/JumpCloud.psd1” -force`
4. Test `Get-JcUser`, `Set-JcUser`, `New-JcUser` to ensure all cases work as intended

## New-JcUser
```PowerShell
#Create a new user with the state STAGED
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -state "STAGED"
#Should return STAGED
$NewUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```

```PowerShell
#Create a new user with the state SUSPENDED
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -state "SUSPENDED"
#Should return SUSPENDED
$NewUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```

```PowerShell
#Create a new user with the state SUSPENDED
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -suspended $true
#Should return SUSPENDED
$NewUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```

```PowerShell
#Create a new user with the state ACTIVATED
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -state "ACTIVATED"
#Should return ACTIVATED
$NewUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```

## Get-JcUser

```PowerShell 
#Should return users with the STAGED state
Get-JcUser -state "STAGED"
```

```PowerShell 
#Should return users with the ACTIVATED state
Get-JcUser -state "ACTIVATED" 
```

```PowerShell 
#Should return users with the SUSPENDED state
Get-JcUser -state "SUSPENDED" 
```

## Set-JcUser by Pipeline
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123!
#Pipe in the user object to Set-JcUser
$UpdatedUser = $NewUser | Set-JcUser -state "SUSPENDED"
#Should return SUSPENDED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -state "SUSPENDED"
#Pipe in the user object to Set-JcUser
$UpdatedUser = $NewUser | Set-JcUser -state "ACTIVATED"
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -suspended $true
#Pipe in the user object to Set-JcUser
$UpdatedUser = $NewUser | Set-JcUser -state "ACTIVATED"
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -suspended $true
#Pipe in the user object to Set-JcUser
$UpdatedUser = $NewUser | Set-JcUser -suspended $false
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
## Set-JcUser by Username
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123!
#Use -Username in Set-JcUser
$UpdatedUser = Set-JcUser -Username $NewUser.username -state "SUSPENDED"
#Should return SUSPENDED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -state "SUSPENDED"
#Use -Username in Set-JcUser
$UpdatedUser = Set-JcUser -Username $NewUser.username -state "ACTIVATED"
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -suspended $true
#Use -Username in Set-JcUser
$UpdatedUser = Set-JcUser -Username $NewUser.username -state "ACTIVATED"
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -suspended $true
#Use -Username in Set-JcUser
$UpdatedUser = Set-JcUser -Username $NewUser.username -suspended $false
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
## Set-JcUser by ID
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123!
#Use -UserID in Set-JcUser
$UpdatedUser = Set-JcUser -UserID $NewUser.id -state "SUSPENDED"
#Should return SUSPENDED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -state "SUSPENDED"
#Use -UserID in Set-JcUser
$UpdatedUser = Set-JcUser -UserID $NewUser.id -state "ACTIVATED"
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -suspended $true
#Use -UserID in Set-JcUser
$UpdatedUser = Set-JcUser -UserID $NewUser.id -state "ACTIVATED"
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
```PowerShell
#Create a new user
$NewUser = New-JcUser -firstname "User" -lastname "Test" -username "user.test" -email "user.test@deleteme.com" -password Test@123! -suspended $true
#Use -UserID in Set-JcUser
$UpdatedUser = Set-JcUser -UserID $NewUser.id -suspended $false
#Should return ACTIVATED
$UpdatedUser.state
#Remove the user that was just created
Remove-JcUser -UserId $NewUser.id
```
